odoo/addons/base/models/ir_cron.py

```diff
@@ -94,14 +94,18 @@ def _process_jobs(cls, db_name):
             threading.current_thread().dbname = db_name
             with db.cursor() as cron_cr:
                 cls._check_version(cron_cr)
-                jobs = cls._get_all_ready_jobs(cron_cr)
+
+                cron_cr.execute("SELECT now() at time zone 'UTC'")
+                now = cron_cr.fetchone()[0]
+
+                jobs = cls._get_all_ready_jobs(cron_cr, now)
                 if not jobs:
                     return
                 cls._check_modules_state(cron_cr, jobs)
                 job_ids = tuple([job['id'] for job in jobs])
 
                 while True:
-                    job = cls._acquire_one_job(cron_cr, job_ids)
+                    job = cls._acquire_one_job(cron_cr, job_ids, now)
```

Comment:
If I understand correctly, this call is the crux of the problem, because it will be called inside different transactions, with a sliding notion of `now()`, and may therefore call *n* times the same job, instead of calling 1 time *n* different jobs?

Then I have 1 question: why do we need to have `_acquire_one_job()` perform the time-based selection as well? All it needs to do is to try to lock a single job, 1 from the `job_ids`, no?

This leads to a simpler solution I think:
- replace the query inside `_acquire_one_job()` with a simple `SELECT * FROM ir_cron WHERE id in %job_ids LIMIT 1 FOR UPDATE SKIP LOCKED`. Since it returns the selected job, the called (`_process_jobs`) can pop it from the list of `job_ids` for each successive call, making sure it won't call the same one twice.
- conceptually you could even have `_process_jobs` call `_acquire_one_job` with a single job_id at a time, you'd just need to keep looping instead of breaking on the first miss, in case a job is locked already.

Or am I missing something?

Reply:
great idea !

Reply:
Or is it... when there are multiple cron workers, we want that a specific cron is only run by a single worker. the `FOR UPDATE` ensures a cron is not run *at the same time* by two workers, the `nextcall < now()` ensures a cron is not run *at different times* by two workers. So the `nextcall < now()` bit is important.